text/0041-fixed-point.md

```diff
@@ -0,0 +1,104 @@
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: [amaranth-lang/rfcs#41](https://github.com/amaranth-lang/rfcs/pull/41)
+- Amaranth Issue: [amaranth-lang/amaranth#0000](https://github.com/amaranth-lang/amaranth/issues/0000)
+
+# Fixed point types
+
+## Summary
+[summary]: #summary
+
+Add fixed point types to Amaranth.
+
+## Motivation
+[motivation]: #motivation
+
+Fractional values in hardware are usually represented as some form of fixed point value.
+Without a first class fixed point type, the user has to manually manage how the value needs to be shifted to be represented as an integer and keep track of how that interacts with arithmetic operations.
+
+A fixed point type would encode and keep track of the precision through arithmetic operations, as well as provide standard operations for converting values to and from fixed point representations with correct rounding.
+
+## Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+TODO
+
+## Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This RFC proposes a library addition `amaranth.lib.fixedpoint` with the following contents:
+
+`FixedPoint` is a `ShapeCastable` subclass.
+The following operations are defined on it:
+
+- `FixedPoint(f_width, /, *, signed)`: Create a `FixedPoint` with `f_width` fractional bits.
+- `FixedPoint(i_width, f_width, /, *, signed)`: Create a `FixedPoint` with `i_width` integer bits and `f_width` fractional bits.
+- `FixedPoint.cast(shape)`: Cast `shape` to a `FixedPoint` instance.
```

Comment:
What is the semantics of this operation?

Reply:
I just updated it to `.cast(shape, f_width=0)` locally. The idea is to determine `i_width` automatically, so `FixedPoint.cast(unsigned(8))` would result in `UQ(8)` and `FixedPoint.cast(unsigned(8), f_width = 4)` would result in `UQ(4, 4)`.

Reply:
Why `UQ(4,4)`?

Reply:
`FixedPoint.cast()` is a building block for `FixedPointValue.cast()`. The idea is to be able to say «here's a value with n fractional bits, please turn it into an appropriately sized `FixedPointValue`». An `unsigned(8)` with four fractional bits would have four integer bits left and thus cast to `UQ(4, 4)`.

Reply:
This API seems confusing and difficult to use. Should we expose it at all?